PGD re-edit batch 8 - reference (#5852)
EnterpriseDB/docs: product_docs/docs/pgd/5/reference/conflicts.mdx

indexdepth: 2
---

> ## Conflict detection

Author: Single pound sign is used only at top level.

-   `update_pkey_exists` &mdash; An incoming update has modified the `PRIMARY KEY` to a value that already exists on the node that's applying the change.
-   `multiple_unique_conflicts` &mdash; The incoming row conflicts with multiple UNIQUE constraints/indexes in the target table.

> -   `delete_recently_updated` &mdash; An incoming delete with an older commit timestamp than the most recent update of the row on the current node or when using [row version conflict detection](../consistency/conflicts/03_conflict_detection/#row-version-conflict-detection).

Author: A verb is missing here. Is it the word "conflicts"?

> <!-- Missing verb>

Author: ```suggestion
```

-   `delete_missing` &mdash; An incoming delete is trying to remove a row that doesn't exist.
-   `target_column_missing` &mdash; The target table is missing one or more columns present in the incoming row.
-   `source_column_missing` &mdash; The incoming row is missing one or more columns that are present in the target table.
-   `target_table_missing` &mdash; The target table is missing.
-   `apply_error_ddl` &mdash; An error was thrown by Postgres when applying a replicated DDL command.

## Conflict resolution

Most conflicts can be resolved automatically. PGD defaults to a last-update-wins mechanism or, more accurately, the `update_if_newer` conflict resolver. This mechanism retains the most recently inserted or changed row of the two conflicting ones based on the same commit timestamps used for conflict detection. The behavior in certain corner-case scenarios depends on the settings used for `bdr.create_node_group` and, alternatively, for `bdr.alter_node_group`.

> PGD lets you override the default behavior of conflict resolution by using the following function.

Author: Instead of saying "following function," should this say "conflict resolvers"? There's no function that follows.

Reviewer: ```suggestion
PGD lets you override the default behavior of conflict resolution by using the `bdr.alter_node_set_conflict_resolver()` function.
```

> <!-- There is no function that follows >

Author: ```suggestion
```

### List of conflict resolvers

Several conflict resolvers are available in PGD, with differing coverages of the conflict types they can handle:

-   `error` &mdash; Throws error and stops replication. Can be used for any conflict type.
-   `skip` &mdash; Skips processing the remote change and continues replication with the next change. Can be used for `insert_exists`, `update_differing`, `update_origin_change`, `update_missing`, `update_recently_deleted`, `update_pkey_exists`, `delete_recently_updated`, `delete_missing`, `target_table_missing`, `target_column_missing`, and `source_column_missing` conflict types.
-   `skip_if_recently_dropped` &mdash; Skips the remote change if it's for a table that doesn't exist downstream because it was recently (within one day) dropped on the downstream. Throw an error otherwise. Can be used for the `target_table_missing` conflict type. `skip_if_recently_dropped` conflict

> resolver can pose challenges if a table with the same name is re-created shortly after it's dropped. In that case, one of the nodes might see the DMLs on the re-created table before it sees the DDL to re-create the table. It then incorrectly skips the remote data, assuming that the table is recently dropped, and causes data loss. We therefore recommend that, when using this conflict resolver, you don't reuse the object names immediately after they're dropped.

Author: I had a hard time following the last sentence. Is this edit correct?

Reviewer: Looks fine.

> <!--Check this edit>

Author: ```suggestion
```
